components/motor/ulnstepper/28byj-48.go

```diff
@@ -150,58 +151,70 @@ func new28byj(
 	}
 	m.in4 = in4
 
+	m.doRun()
+
 	return m, nil
 }
 
 // struct is named after the controler uln28byj.
 type uln28byj struct {
 	resource.Named
 	resource.AlwaysRebuild
-	resource.TriviallyCloseable
 	theBoard           board.Board
 	ticksPerRotation   int
 	in1, in2, in3, in4 board.GPIOPin
 	logger             logging.Logger
 	motorName          string
 
 	// state
-	lock  sync.Mutex
-	opMgr *operation.SingleOperationManager
+	workers utils.StoppableWorkers
+	lock    sync.RWMutex
+	opMgr   *operation.SingleOperationManager
 
 	stepPosition       int64
 	stepperDelay       time.Duration
 	targetStepPosition int64
+	lastTargetPosition int64
 }
 
 // doRun runs the motor till it reaches target step position.
-func (m *uln28byj) doRun(ctx context.Context) error {
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		default:
-		}
-
-		m.lock.Lock()
+func (m *uln28byj) doRun() {
+	// Spawn a new goroutine to do all the work in the background.
+	m.workers = utils.NewStoppableWorkers(func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
 
-		// This condition cannot be locked for the duration of the loop as
-		// Stop() modifies m.targetStepPosition to interrupt the run
-		if m.stepPosition == m.targetStepPosition {
-			err := m.setPins(ctx, [4]bool{false, false, false, false})
-			if err != nil {
-				return errors.Wrapf(err, "error while disabling motor (%s)", m.motorName)
+			m.lock.Lock()
+			// if the target position hasn't changed, don't do anything
+			if m.targetStepPosition == m.lastTargetPosition {
+				m.lock.Unlock()
+				continue
```

Comment:
Let's try to make this less lock-y in a refactor. I think the non-blocking APIs can set the targets in a lock and defer until they return, they do not block, and then you can explicitly set the targets in a scope with the blocking GoFor API.

Reply:
I can take a look today and if it doesn't seem quick to add here I'll make a follow up?

Reply:
Actually, I don't think it matters if it comes from a blocking or non-blocking call because this just starts `doRun` once and then everything is based on what the current and target positions are. so no locks would persist from any of the motor apis. I can add a defer in `SetPower` tho. I think if we defer unlock in `doRun` everything becomes blocking and stop can't interrupt

Reply:
Then I think this is the wrong design for this driver since it is too easy to change the place for an unlock and cause a hang. We had this problem in gpio motor and had to refactor it back bit by bit to its current form. I think it's okay to address this in this pr rather than a follow-up. We don't need to merge this in immediately.

Reply:
sounds good, I'll structure it like gpio motor instead

Reply:
okay actually I tried something different. now every function, except `GoFor` defers unlock. `GoFor` still locks only to update `m.targetStepPosition` and `m.stepperDelay`. Then in `doRun`, `m.targetStepPosition` is always read using `sync.atomic` since that's the only variable we really care about changing, to know if stop is interrupting a movement call. This might be overcomplicated but I think the other option is adding a `doRun` context and starting `doRun` every time there is a new movement call, like gpio motor does, which is seems more complicated to me. But if you'd prefer that it should be easy to implement so no worries. 

with this implementation, I think it's fine to leave `m.stepPosition` not protected by any locks in `doRun` since we lock everywhere else we update it, but we could alternatively read `m.stepPosition` with `sync.atomic` as well.

Reply:
hmm maybe this won't work, there's some test race failures

Reply:
okays so got it to a working state with no data race, but I now think it will be better to implement as gpio motor does. I thought that lock and atomic could be used together, meaning I could use `atomic.Store` and `atomic.Load` _only_ when outside of a regular lock, but it seems like if you use atomic anywhere, you have to use it everywhere on that variable. Not 100% of the details I'll have to read more about it, but in the meantime I will refactor this to be structured the way gpio motor is